File: tests/rosetta/transpiler/FS/2048.fs
// Generated 2025-07-25 12:29 +0700

exception Break
exception Continue

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
type Board = {
    cells: int array array
}
type SpawnResult = {
    board: Board
    full: bool
}
type SlideResult = {
    row: int array
    gain: int
}
type MoveResult = {
    board: Board
    score: int
    moved: bool
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
open System

let SIZE: int = 4
let rec newBoard () =
    let mutable __ret : Board = Unchecked.defaultof<Board>
    try
        let mutable b: int array array = [||]
        let mutable y: int = 0
        while y < SIZE do
            let mutable row: int array = [||]
            let mutable x: int = 0
            while x < SIZE do
                row <- Array.append row [|0|]
                x <- x + 1
            b <- Array.append b [|row|]
            y <- y + 1
        __ret <- { cells = b }
        raise Return
        __ret
    with
        | Return -> __ret
let rec spawnTile (b: Board) =
    let mutable __ret : SpawnResult = Unchecked.defaultof<SpawnResult>
    let mutable b = b
    try
        let mutable grid: int array array = b.cells
        let mutable empty: int array array = [||]
        let mutable y: int = 0
        while y < SIZE do
            let mutable x: int = 0
            while x < SIZE do
                if (grid.[y].[x]) = 0 then
                    empty <- Array.append empty [|[|x; y|]|]
                x <- x + 1
            y <- y + 1
        if (Array.length empty) = 0 then
            __ret <- { board = b; full = true }
            raise Return
        let mutable idx = (_now()) % (Array.length empty)
        let cell = empty.[idx]
        let mutable ``val``: int = 4
        if ((_now()) % 10) < 9 then
            ``val`` <- 2
        grid.[cell.[1]].[cell.[0]] <- ``val``
        __ret <- { board = { cells = grid }; full = (Array.length empty) = 1 }
        raise Return
        __ret
    with
        | Return -> __ret
let rec pad (n: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable n = n
    try
        let mutable s: string = string n
        let mutable pad: int = 4 - (String.length s)
        let mutable i: int = 0
        let mutable out: string = ""
        while i < pad do
            out <- out + " "
            i <- i + 1
        __ret <- out + s
        raise Return
        __ret
    with
        | Return -> __ret
let rec draw (b: Board) (score: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable score = score
    try
        printfn "%s" ("Score: " + (string score))
        let mutable y: int = 0
        while y < SIZE do
            printfn "%s" "+----+----+----+----+"
            let mutable line: string = "|"
            let mutable x: int = 0
            while x < SIZE do
                let mutable v = b.cells.[y].[x]
                if v = 0 then
                    line <- line + "    |"
                else
                    line <- (line + (pad v)) + "|"
                x <- x + 1
            printfn "%s" line
            y <- y + 1
        printfn "%s" "+----+----+----+----+"
        printfn "%s" "W=Up S=Down A=Left D=Right Q=Quit"
        __ret
    with
        | Return -> __ret
let rec reverseRow (r: int array) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable r = r
    try
        let mutable out: int array = [||]
        let mutable i = (Array.length r) - 1
        while i >= 0 do
            out <- Array.append out [|r.[i]|]
            i <- i - 1
        __ret <- out
        raise Return
        __ret
    with
        | Return -> __ret
let rec slideLeft (row: int array) =
    let mutable __ret : SlideResult = Unchecked.defaultof<SlideResult>
    let mutable row = row
    try
        let mutable xs: int array = [||]
        let mutable i: int = 0
        while i < (Array.length row) do
            if (row.[i]) <> 0 then
                xs <- Array.append xs [|row.[i]|]
            i <- i + 1
        let mutable res: int array = [||]
        let mutable gain: int = 0
        i <- 0
        while i < (Array.length xs) do
            if ((i + 1) < (Array.length xs)) && ((xs.[i]) = (xs.[i + 1])) then
                let v = (xs.[i]) * 2
                gain <- gain + v
                res <- Array.append res [|v|]
                i <- i + 2
            else
                res <- Array.append res [|xs.[i]|]
                i <- i + 1
        while (Array.length res) < SIZE do
            res <- Array.append res [|0|]
        __ret <- { row = res; gain = gain }
        raise Return
        __ret
    with
        | Return -> __ret
let rec moveLeft (b: Board) (score: int) =
    let mutable __ret : MoveResult = Unchecked.defaultof<MoveResult>
    let mutable b = b
    let mutable score = score
    try
        let mutable grid: int array array = b.cells
        let mutable moved: bool = false
        let mutable y: int = 0
        while y < SIZE do
            let r = slideLeft (grid.[y])
            let ``new`` = r.row
            score <- score + (r.gain)
            let mutable x: int = 0
            while x < SIZE do
                if (grid.[y].[x]) <> (``new``.[x]) then
                    moved <- true
                grid.[y].[x] <- ``new``.[x]
                x <- x + 1
            y <- y + 1
        __ret <- { board = { cells = grid }; score = score; moved = moved }
        raise Return
        __ret
    with
        | Return -> __ret
let rec moveRight (b: Board) (score: int) =
    let mutable __ret : MoveResult = Unchecked.defaultof<MoveResult>
    let mutable b = b
    let mutable score = score
    try
        let mutable grid: int array array = b.cells
        let mutable moved: bool = false
        let mutable y: int = 0
        while y < SIZE do
            let mutable rev = reverseRow (grid.[y])
            let r = slideLeft rev
            rev <- r.row
            score <- score + (r.gain)
            rev <- reverseRow rev
            let mutable x: int = 0
            while x < SIZE do
                if (grid.[y].[x]) <> (rev.[x]) then
                    moved <- true
                grid.[y].[x] <- rev.[x]
                x <- x + 1
            y <- y + 1
        __ret <- { board = { cells = grid }; score = score; moved = moved }
        raise Return
        __ret
    with
        | Return -> __ret
let rec getCol (b: Board) (x: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable b = b
    let mutable x = x
    try
        let mutable col: int array = [||]
        let mutable y: int = 0
        while y < SIZE do
            col <- Array.append col [|b.cells.[y].[x]|]
            y <- y + 1
        __ret <- col
        raise Return
        __ret
    with
        | Return -> __ret
let rec setCol (b: Board) (x: int) (col: int array) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable x = x
    let mutable col = col
    try
        let mutable rows: int array array = b.cells
        let mutable y: int = 0
        while y < SIZE do
            let mutable row = rows.[y]
            row.[x] <- col.[y]
            rows.[y] <- row
            y <- y + 1
        b <- { b with cells = rows }
        __ret
    with
        | Return -> __ret
let rec moveUp (b: Board) (score: int) =
    let mutable __ret : MoveResult = Unchecked.defaultof<MoveResult>
    let mutable b = b
    let mutable score = score
    try
        let mutable grid: int array array = b.cells
        let mutable moved: bool = false
        let mutable x: int = 0
        while x < SIZE do
            let mutable col = getCol b x
            let r = slideLeft col
            let ``new`` = r.row
            score <- score + (r.gain)
            let mutable y: int = 0
            while y < SIZE do
                if (grid.[y].[x]) <> (``new``.[y]) then
                    moved <- true
                grid.[y].[x] <- ``new``.[y]
                y <- y + 1
            x <- x + 1
        __ret <- { board = { cells = grid }; score = score; moved = moved }
        raise Return
        __ret
    with
        | Return -> __ret
let rec moveDown (b: Board) (score: int) =
    let mutable __ret : MoveResult = Unchecked.defaultof<MoveResult>
    let mutable b = b
    let mutable score = score
    try
        let mutable grid: int array array = b.cells
        let mutable moved: bool = false
        let mutable x: int = 0
        while x < SIZE do
            let mutable col = reverseRow (getCol b x)
            let r = slideLeft col
            col <- r.row
            score <- score + (r.gain)
            col <- reverseRow col
            let mutable y: int = 0
            while y < SIZE do
                if (grid.[y].[x]) <> (col.[y]) then
                    moved <- true
                grid.[y].[x] <- col.[y]
                y <- y + 1
            x <- x + 1
        __ret <- { board = { cells = grid }; score = score; moved = moved }
        raise Return
        __ret
    with
        | Return -> __ret
let rec hasMoves (b: Board) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable b = b
    try
        let mutable y: int = 0
        while y < SIZE do
            let mutable x: int = 0
            while x < SIZE do
                if (b.cells.[y].[x]) = 0 then
                    __ret <- true
                    raise Return
                if ((x + 1) < SIZE) && ((b.cells.[y].[x]) = (b.cells.[y].[x + 1])) then
                    __ret <- true
                    raise Return
                if ((y + 1) < SIZE) && ((b.cells.[y].[x]) = (b.cells.[y + 1].[x])) then
                    __ret <- true
                    raise Return
                x <- x + 1
            y <- y + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
let rec has2048 (b: Board) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable b = b
    try
        let mutable y: int = 0
        while y < SIZE do
            let mutable x: int = 0
            while x < SIZE do
                if (b.cells.[y].[x]) >= 2048 then
                    __ret <- true
                    raise Return
                x <- x + 1
            y <- y + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
let mutable board: Board = newBoard()
let mutable r = spawnTile board
board <- r.board
let mutable full = r.full
r <- spawnTile board
board <- r.board
full <- r.full
let mutable score: int = 0
draw board score
try
    while true do
        printfn "%s" "Move: "
        let cmd: string = System.Console.ReadLine()
        let mutable moved: bool = false
        if (cmd = "a") || (cmd = "A") then
            let m = moveLeft board score
            board <- m.board
            score <- m.score
            moved <- m.moved
        if (cmd = "d") || (cmd = "D") then
            let m = moveRight board score
            board <- m.board
            score <- m.score
            moved <- m.moved
        if (cmd = "w") || (cmd = "W") then
            let m = moveUp board score
            board <- m.board
            score <- m.score
            moved <- m.moved
        if (cmd = "s") || (cmd = "S") then
            let m = moveDown board score
            board <- m.board
            score <- m.score
            moved <- m.moved
        if (cmd = "q") || (cmd = "Q") then
            raise Break
        if moved then
            let r2 = spawnTile board
            board <- r2.board
            full <- r2.full
            if full && (not (hasMoves board)) then
                draw board score
                printfn "%s" "Game Over"
                raise Break
        draw board score
        if has2048 board then
            printfn "%s" "You win!"
            raise Break
        if not (hasMoves board) then
            printfn "%s" "Game Over"
            raise Break
with
| Break -> ()
| Continue -> ()
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/rosetta/transpiler/Pascal/2048.pas
{$mode objfpc}
program Main;
uses SysUtils;
type IntArray = array of integer;
type IntArrayArray = array of IntArray;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
function _input(): string;
var s: string;
begin
  if EOF(Input) then s := '' else ReadLn(s);
  _input := s;
end;
type Board = record
  cells: array of IntArray;
end;
type SpawnResult = record
  board: Board;
  full: boolean;
end;
type SlideResult = record
  row: array of integer;
  gain: integer;
end;
type MoveResult = record
  board: Board;
  score: integer;
  moved: boolean;
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  SIZE: integer;
  newBoard_b: array of IntArray;
  newBoard_y: integer;
  newBoard_row: array of integer;
  newBoard_x: integer;
  spawnTile_grid: array of IntArray;
  spawnTile_empty: array of IntArray;
  spawnTile_y: integer;
  spawnTile_x: integer;
  spawnTile_idx: integer;
  spawnTile_cell: IntArray;
  spawnTile_val: integer;
  pad_s: string;
  pad_pad_var: integer;
  pad_i: integer;
  pad_out: string;
  draw_y: integer;
  draw_line: string;
  draw_x: integer;
  draw_v: integer;
  reverseRow_out: array of integer;
  reverseRow_i: integer;
  slideLeft_xs: array of integer;
  slideLeft_i: integer;
  slideLeft_res: array of integer;
  slideLeft_gain: integer;
  slideLeft_v: integer;
  moveLeft_grid: array of IntArray;
  moveLeft_moved: boolean;
  moveLeft_y: integer;
  moveLeft_r: SlideResult;
  moveLeft_new: array of integer;
  moveLeft_x: integer;
  moveRight_grid: array of IntArray;
  moveRight_moved: boolean;
  moveRight_y: integer;
  moveRight_rev: IntArray;
  moveRight_r: SlideResult;
  moveRight_x: integer;
  getCol_col: array of integer;
  getCol_y: integer;
  setCol_rows: array of IntArray;
  setCol_y: integer;
  setCol_row: IntArray;
  moveUp_grid: array of IntArray;
  moveUp_moved: boolean;
  moveUp_x: integer;
  moveUp_col: IntArray;
  moveUp_r: SlideResult;
  moveUp_new: array of integer;
  moveUp_y: integer;
  moveDown_grid: array of IntArray;
  moveDown_moved: boolean;
  moveDown_x: integer;
  moveDown_col: IntArray;
  moveDown_r: SlideResult;
  moveDown_y: integer;
  hasMoves_y: integer;
  hasMoves_x: integer;
  has2048_y: integer;
  has2048_x: integer;
  board_var: Board;
  r: SpawnResult;
  full: boolean;
  score: integer;
  cmd: string;
  moved: boolean;
  m: MoveResult;
  r2: SpawnResult;
function makeMoveResult(board: Board; score: integer; moved: boolean): MoveResult;
begin
  Result.board := board;
  Result.score := score;
  Result.moved := moved;
end;
function makeSlideResult(row: IntArray; gain: integer): SlideResult;
begin
  Result.row := row;
  Result.gain := gain;
end;
function makeSpawnResult(board: Board; full: boolean): SpawnResult;
begin
  Result.board := board;
  Result.full := full;
end;
function makeBoard(cells: IntArrayArray): Board;
begin
  Result.cells := cells;
end;
function newBoard(): Board;
begin
  newBoard_b := [];
  newBoard_y := 0;
  while newBoard_y < SIZE do begin
  newBoard_row := [];
  newBoard_x := 0;
  while newBoard_x < SIZE do begin
  newBoard_row := concat(newBoard_row, [0]);
  newBoard_x := newBoard_x + 1;
end;
  newBoard_b := concat(newBoard_b, [newBoard_row]);
  newBoard_y := newBoard_y + 1;
end;
  exit(makeBoard(newBoard_b));
end;
function spawnTile(b: Board): SpawnResult;
begin
  spawnTile_grid := b.cells;
  spawnTile_empty := [];
  spawnTile_y := 0;
  while spawnTile_y < SIZE do begin
  spawnTile_x := 0;
  while spawnTile_x < SIZE do begin
  if spawnTile_grid[spawnTile_y][spawnTile_x] = 0 then begin
  spawnTile_empty := concat(spawnTile_empty, [[spawnTile_x, spawnTile_y]]);
end;
  spawnTile_x := spawnTile_x + 1;
end;
  spawnTile_y := spawnTile_y + 1;
end;
  if Length(spawnTile_empty) = 0 then begin
  exit(makeSpawnResult(b, true));
end;
  spawnTile_idx := _now() mod Length(spawnTile_empty);
  spawnTile_cell := spawnTile_empty[spawnTile_idx];
  spawnTile_val := 4;
  if (_now() mod 10) < 9 then begin
  spawnTile_val := 2;
end;
  spawnTile_grid[spawnTile_cell[1]][spawnTile_cell[0]] := spawnTile_val;
  exit(makeSpawnResult(makeBoard(spawnTile_grid), Length(spawnTile_empty) = 1));
end;
function pad(n: integer): string;
begin
  pad_s := IntToStr(n);
  pad_pad_var := 4 - Length(pad_s);
  pad_i := 0;
  pad_out := '';
  while pad_i < pad_pad_var do begin
  pad_out := pad_out + ' ';
  pad_i := pad_i + 1;
end;
  exit(pad_out + pad_s);
end;
procedure draw(b: Board; score: integer);
begin
  writeln('Score: ' + IntToStr(score));
  draw_y := 0;
  while draw_y < SIZE do begin
  writeln('+----+----+----+----+');
  draw_line := '|';
  draw_x := 0;
  while draw_x < SIZE do begin
  draw_v := b.cells[draw_y][draw_x];
  if draw_v = 0 then begin
  draw_line := draw_line + '    |';
end else begin
  draw_line := (draw_line + pad(draw_v)) + '|';
end;
  draw_x := draw_x + 1;
end;
  writeln(draw_line);
  draw_y := draw_y + 1;
end;
  writeln('+----+----+----+----+');
  writeln('W=Up S=Down A=Left D=Right Q=Quit');
end;
function reverseRow(r: IntArray): IntArray;
begin
  reverseRow_out := [];
  reverseRow_i := Length(r) - 1;
  while reverseRow_i >= 0 do begin
  reverseRow_out := concat(reverseRow_out, [r[reverseRow_i]]);
  reverseRow_i := reverseRow_i - 1;
end;
  exit(reverseRow_out);
end;
function slideLeft(row: IntArray): SlideResult;
begin
  slideLeft_xs := [];
  slideLeft_i := 0;
  while slideLeft_i < Length(row) do begin
  if row[slideLeft_i] <> 0 then begin
  slideLeft_xs := concat(slideLeft_xs, [row[slideLeft_i]]);
end;
  slideLeft_i := slideLeft_i + 1;
end;
  slideLeft_res := [];
  slideLeft_gain := 0;
  slideLeft_i := 0;
  while slideLeft_i < Length(slideLeft_xs) do begin
  if ((slideLeft_i + 1) < Length(slideLeft_xs)) and (slideLeft_xs[slideLeft_i] = slideLeft_xs[slideLeft_i + 1]) then begin
  slideLeft_v := slideLeft_xs[slideLeft_i] * 2;
  slideLeft_gain := slideLeft_gain + slideLeft_v;
  slideLeft_res := concat(slideLeft_res, [slideLeft_v]);
  slideLeft_i := slideLeft_i + 2;
end else begin
  slideLeft_res := concat(slideLeft_res, [slideLeft_xs[slideLeft_i]]);
  slideLeft_i := slideLeft_i + 1;
end;
end;
  while Length(slideLeft_res) < SIZE do begin
  slideLeft_res := concat(slideLeft_res, [0]);
end;
  exit(makeSlideResult(slideLeft_res, slideLeft_gain));
end;
function moveLeft(b: Board; score: integer): MoveResult;
begin
  moveLeft_grid := b.cells;
  moveLeft_moved := false;
  moveLeft_y := 0;
  while moveLeft_y < SIZE do begin
  moveLeft_r := slideLeft(moveLeft_grid[moveLeft_y]);
  moveLeft_new := moveLeft_r.row;
  score := score + moveLeft_r.gain;
  moveLeft_x := 0;
  while moveLeft_x < SIZE do begin
  if moveLeft_grid[moveLeft_y][moveLeft_x] <> moveLeft_new[moveLeft_x] then begin
  moveLeft_moved := true;
end;
  moveLeft_grid[moveLeft_y][moveLeft_x] := moveLeft_new[moveLeft_x];
  moveLeft_x := moveLeft_x + 1;
end;
  moveLeft_y := moveLeft_y + 1;
end;
  exit(makeMoveResult(makeBoard(moveLeft_grid), score, moveLeft_moved));
end;
function moveRight(b: Board; score: integer): MoveResult;
begin
  moveRight_grid := b.cells;
  moveRight_moved := false;
  moveRight_y := 0;
  while moveRight_y < SIZE do begin
  moveRight_rev := reverseRow(moveRight_grid[moveRight_y]);
  moveRight_r := slideLeft(moveRight_rev);
  moveRight_rev := moveRight_r.row;
  score := score + moveRight_r.gain;
  moveRight_rev := reverseRow(moveRight_rev);
  moveRight_x := 0;
  while moveRight_x < SIZE do begin
  if moveRight_grid[moveRight_y][moveRight_x] <> moveRight_rev[moveRight_x] then begin
  moveRight_moved := true;
end;
  moveRight_grid[moveRight_y][moveRight_x] := moveRight_rev[moveRight_x];
  moveRight_x := moveRight_x + 1;
end;
  moveRight_y := moveRight_y + 1;
end;
  exit(makeMoveResult(makeBoard(moveRight_grid), score, moveRight_moved));
end;
function getCol(b: Board; x: integer): IntArray;
begin
  getCol_col := [];
  getCol_y := 0;
  while getCol_y < SIZE do begin
  getCol_col := concat(getCol_col, [b.cells[getCol_y][x]]);
  getCol_y := getCol_y + 1;
end;
  exit(getCol_col);
end;
procedure setCol(b: Board; x: integer; col: IntArray);
begin
  setCol_rows := b.cells;
  setCol_y := 0;
  while setCol_y < SIZE do begin
  setCol_row := setCol_rows[setCol_y];
  setCol_row[x] := col[setCol_y];
  setCol_rows[setCol_y] := setCol_row;
  setCol_y := setCol_y + 1;
end;
  b.cells := setCol_rows;
end;
function moveUp(b: Board; score: integer): MoveResult;
begin
  moveUp_grid := b.cells;
  moveUp_moved := false;
  moveUp_x := 0;
  while moveUp_x < SIZE do begin
  moveUp_col := getCol(b, moveUp_x);
  moveUp_r := slideLeft(moveUp_col);
  moveUp_new := moveUp_r.row;
  score := score + moveUp_r.gain;
  moveUp_y := 0;
  while moveUp_y < SIZE do begin
  if moveUp_grid[moveUp_y][moveUp_x] <> moveUp_new[moveUp_y] then begin
  moveUp_moved := true;
end;
  moveUp_grid[moveUp_y][moveUp_x] := moveUp_new[moveUp_y];
  moveUp_y := moveUp_y + 1;
end;
  moveUp_x := moveUp_x + 1;
end;
  exit(makeMoveResult(makeBoard(moveUp_grid), score, moveUp_moved));
end;
function moveDown(b: Board; score: integer): MoveResult;
begin
  moveDown_grid := b.cells;
  moveDown_moved := false;
  moveDown_x := 0;
  while moveDown_x < SIZE do begin
  moveDown_col := reverseRow(getCol(b, moveDown_x));
  moveDown_r := slideLeft(moveDown_col);
  moveDown_col := moveDown_r.row;
  score := score + moveDown_r.gain;
  moveDown_col := reverseRow(moveDown_col);
  moveDown_y := 0;
  while moveDown_y < SIZE do begin
  if moveDown_grid[moveDown_y][moveDown_x] <> moveDown_col[moveDown_y] then begin
  moveDown_moved := true;
end;
  moveDown_grid[moveDown_y][moveDown_x] := moveDown_col[moveDown_y];
  moveDown_y := moveDown_y + 1;
end;
  moveDown_x := moveDown_x + 1;
end;
  exit(makeMoveResult(makeBoard(moveDown_grid), score, moveDown_moved));
end;
function hasMoves(b: Board): boolean;
begin
  hasMoves_y := 0;
  while hasMoves_y < SIZE do begin
  hasMoves_x := 0;
  while hasMoves_x < SIZE do begin
  if b.cells[hasMoves_y][hasMoves_x] = 0 then begin
  exit(true);
end;
  if ((hasMoves_x + 1) < SIZE) and (b.cells[hasMoves_y][hasMoves_x] = b.cells[hasMoves_y][hasMoves_x + 1]) then begin
  exit(true);
end;
  if ((hasMoves_y + 1) < SIZE) and (b.cells[hasMoves_y][hasMoves_x] = b.cells[hasMoves_y + 1][hasMoves_x]) then begin
  exit(true);
end;
  hasMoves_x := hasMoves_x + 1;
end;
  hasMoves_y := hasMoves_y + 1;
end;
  exit(false);
end;
function has2048(b: Board): boolean;
begin
  has2048_y := 0;
  while has2048_y < SIZE do begin
  has2048_x := 0;
  while has2048_x < SIZE do begin
  if b.cells[has2048_y][has2048_x] >= 2048 then begin
  exit(true);
end;
  has2048_x := has2048_x + 1;
end;
  has2048_y := has2048_y + 1;
end;
  exit(false);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _now();
  SIZE := 4;
  board_var := newBoard();
  r := spawnTile(board_var);
  board_var := r.board;
  full := r.full;
  r := spawnTile(board_var);
  board_var := r.board;
  full := r.full;
  score := 0;
  draw(board_var, score);
  while true do begin
  writeln('Move: ');
  cmd := _input();
  moved := false;
  if (cmd = 'a') or (cmd = 'A') then begin
  m := moveLeft(board_var, score);
  board_var := m.board;
  score := m.score;
  moved := m.moved;
end;
  if (cmd = 'd') or (cmd = 'D') then begin
  m := moveRight(board_var, score);
  board_var := m.board;
  score := m.score;
  moved := m.moved;
end;
  if (cmd = 'w') or (cmd = 'W') then begin
  m := moveUp(board_var, score);
  board_var := m.board;
  score := m.score;
  moved := m.moved;
end;
  if (cmd = 's') or (cmd = 'S') then begin
  m := moveDown(board_var, score);
  board_var := m.board;
  score := m.score;
  moved := m.moved;
end;
  if (cmd = 'q') or (cmd = 'Q') then begin
  break;
end;
  if moved then begin
  r2 := spawnTile(board_var);
  board_var := r2.board;
  full := r2.full;
  if full and not hasMoves(board_var) then begin
  draw(board_var, score);
  writeln('Game Over');
  break;
end;
end;
  draw(board_var, score);
  if has2048(board_var) then begin
  writeln('You win!');
  break;
end;
  if not hasMoves(board_var) then begin
  writeln('Game Over');
  break;
end;
end;
  Sleep(1);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
